src/python/grpcio/grpc/_observability.py

```diff
@@ -0,0 +1,125 @@
+import threading
+from typing import Any, Generic, Optional, TypeVar
+
+from grpc._cython import cygrpc as _cygrpc
+
+_LOGGER = logging.getLogger(__name__)
+
+_channel = Any  # _channel.py imports this module.
+ClientCallTracerCapsule = TypeVar('ClientCallTracerCapsule')
+ServerCallTracerFactoryCapsule = TypeVar('ServerCallTracerFactoryCapsule')
+
+_plugin_lock: threading.RLock = threading.RLock()
+_OBSERVABILITY_PLUGIN: Optional[_ObservabilityPlugin] = None  # pylint: disable=used-before-assignment
+
+
+class _ObservabilityPlugin(Generic[ClientCallTracerCapsule,
+                                   ServerCallTracerFactoryCapsule],
+                           metaclass=abc.ABCMeta):
+    """
+    Note: Any future methods added to this interface cannot have the @abc.abstractmethod annotation.
+    """
+    _tracing_enabled: bool = False
+    _stats_enabled: bool = False
+
+    @abc.abstractmethod
+    def create_client_call_tracer(
+            self, method_name: bytes) -> ClientCallTracerCapsule:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def delete_client_call_tracer(
+            self, client_call_tracer: ClientCallTracerCapsule) -> None:
+        # delte client call tracer have to be called on o11y package side.
+        # Call it for both segregated and integrated call (`_process_integrated_call_tag`)
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def save_trace_context(self, trace_id: str, span_id: str,
+                           is_sampled: bool) -> None:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def create_server_call_tracer_factory(
+            self) -> ServerCallTracerFactoryCapsule:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def record_rpc_latency(self, method: str, rpc_latency: float,
+                           status_code: Any) -> None:
+        raise NotImplementedError()
+
+    def set_tracing(self, enable: bool) -> None:
+        self._tracing_enabled = enable
+
+    def set_stats(self, enable: bool) -> None:
+        self._stats_enabled = enable
+
+    @property
+    def tracing_enabled(self) -> bool:
+        return self._tracing_enabled
+
+    @property
+    def stats_enabled(self) -> bool:
+        return self._stats_enabled
+
+    @property
+    def observability_enabled(self) -> bool:
+        return self.tracing_enabled or self.stats_enabled
+
+
+@contextlib.contextmanager
+def get_plugin() -> Optional[_ObservabilityPlugin]:
+    with _plugin_lock:
+        yield _OBSERVABILITY_PLUGIN
+
+
+def set_plugin(observability_plugin: Optional[_ObservabilityPlugin]) -> None:
+    global _OBSERVABILITY_PLUGIN  # pylint: disable=global-statement
+    with _plugin_lock:
+        if observability_plugin and _OBSERVABILITY_PLUGIN:
+            raise ValueError("observability_plugin was already set!")
+        _OBSERVABILITY_PLUGIN = observability_plugin
+
+
+def _observability_init(observability_plugin: _ObservabilityPlugin) -> None:
```

Comment:
What I'm currently seeing is:

- When entering the o11y context manager, so the `observability` module calls `_observability_init`
  - `_observability_init` calls `set_plugin`
  - `_observability_init` calls `set_server_call_tracer_factory`
- When exiting the o11y context manager, the `observability` module calls `set_plugin(None)`

Some thoughts about this.
- I don't see parallelism between the calls that the o11y context manager makes. For example, `set_plugin(foo)` and `set_plugin(None)` or `observability_init(foo)` and `observability_deinit()`
- In the current setup, I don't see us ever cleaning up the server call tracer factory.
- The name scheme here is confusing. `set_plugin` and `_observability_init` are both protected APIs -- ones that should only be used by packages maintained by us but I'm seeing a mixture of underscore patterns. I would suggest letting the underscore of the `_observability` module indicate that people should not touch this and removing the underscores from the functions in this module.

Reply:
Discussed a little bit on this offline, it's more about the naming, I'll change the name to `observability_init` and `observability_deinit` on both `grpc_observability` and `grpc` module.